### ISSUE TITLE
How to implement stateful computations

### ISSUE THREAD

**Issue by evilpoke** [sine-fdn/garble-lang#155]
I plan to implement a private SAT solving algorithm in your framework. The solver of choice is DPLL which essentially does a tree search through the formula, which also requires back-propagation steps.
However seeing your note on the README that 
”All programs written in Garble are deliberately Turing-incomplete (only 
supporting bounded recursion), guaranteeing that they can be compiled to
 circuits using only AND, XOR and NOT gates (without any kind of stateful latches or registers)”
leads me to believe that stateful search algorithms can not be implemented with garble. 

Am I right to assume that Tandem (https://github.com/sine-fdn/tandem?tab=readme-ov-file ) functions as a sort of “wrapper” for Garble allowing the use of mutable datastructures within the circuit execution? In particular I am interested in:
- How does (if present) stateful execution work?
- How does your MPC Engine tandem stand in relation to Garble?
- What does the MPC Engine do, what does Garble do?
- Do you know of any exemplary programs written in your framework which implements some form of tree search algorithm?

**Comment by fkettelhoit** [comment]
First of all, thanks for your interest in Garble!

Regarding the relationship between Tandem and Garble, the most important point is that Tandem is executing Garble programs (or more accurately a circuit generated as output by the Garble compiler) in a way that no private inputs are revealed, in other words using Multi-Party Computation. Tandem does not increase Garble's expressive capabilities in any way, you can do exactly the same in Tandem (when running Garble programs) as you could do if you just simulate them directly using Garble's evaluator.

However, Garble is probably expressive enough to implement your specific algorithm (at least theoretically, you will have to measure whether it is fast enough in practice), because Garble does support stateful execution (you can use mutable variables in Garble and mutate data structures such as arrays), it simply requires all programs to be _bounded_ in their execution. You can use loops in Garble, but you cannot loop infinitely. This restricts how you could work with dynamic structures: You cannot have arrays or lists that grow at runtime, but you can express data structures that have a maximum size.

Long story short, you can probably implement your tree search algorithm in Garble, you just have to specify a maximum size for the tree (and then the Garble program will always take the same time to execute, no matter how big the tree _actually_ is, because Garble programs are just boolean circuits that get fully evaluated no matter what values their input bits take on). You can kind of think as Garble programs as being programs where an algorithm does not have a best case and worst case runtime performance, it will always run with its worst case performance, which is a requirement for using it for cryptographically secure computations.

**Comment by evilpoke** [comment]
thank you for your answer. It was very helpful
